crates/resolver/src/name_server/name_server_pool.rs

```diff
@@ -288,6 +290,8 @@ where
     P: ConnectionProvider<Conn = C> + 'static,
 {
     let mut err = ResolveError::from("No connections available");
+    let mut backoff = Duration::from_millis(100);
+    let mut busy = SmallVec::<[NameServer<C, P>; 2]>::new();
```

Comment:
Could you document the usage of these two things here? I'd like to understand what the intended out come of these when in use are.

Reply:
If I'm understanding correctly, it looks like if the connection is `busy` then we store the nameserver in busy and await the backoff time before trying one of the connections again?

I wonder about fairness here... but that's an optimization we could make later.

Reply:
I added a big fat comment and also changed the code to (a) start from a smaller backoff and (b) use exponential backoff rather than linear. This gives slightly better results on my benchmark, and I was feeling iffy about the linear backoff because as I understand it people almost always use exponential backoff rather than linear.

Reply:
Tweaked this code some more after thinking about fairness. Rather than using `select_ok()`, which I think has some undesirable effects in the face of `ProtoErrorKind::Busy`, this now uses `FuturesUnordered` to drive the underlying futures. This initially caused the concurrent requests to hang, which I fixed in new commit bf48fd7. I also changed ab5925f to derive the `NoRecordsFound` trust directly from the `trusted` bit rather than consulting the `NameServerConfig`.

Reply:
I think those changes look ok. What caused that underlying hang?

Reply:
See my other comment.